### PR TITLE
Add websocket support for 2016 TVs

### LINKED ISSUE #18: Samsung KS8000 support?
Hello, I have a KS8000 Samsung TV and have found that port 55000 is not open. Any ideas on how we could support my TV?

Here's a quick nmap of the running TV:

```
Host is up (0.010s latency).
Not shown: 994 closed ports
PORT     STATE SERVICE
7676/tcp open  imqbrokerd
8000/tcp open  http-alt
8001/tcp open  vcom-tunnel
8002/tcp open  teradataordbms
8080/tcp open  http-proxy
9999/tcp open  abyss
```

### PR DESCRIPTION
**Pull request by nunofgs**
This PR adds support for 2016 Samsung TVs such as the popular KS8000 model. It works great with my TV but I don't know exactly which models it will support!

_Note to reviewers_: my native language is not Python, so I'd very much appreciate any tips or comments regarding my approach.

Closes #18.

### CODE REVIEW

File: samsungctl/__main__.py [hunk 1]
```diff
@@ -96,18 +99,22 @@ def main():
...
-            if args.interactive:
-                logging.getLogger().setLevel(logging.ERROR)
-                interactive.run(remote)
-            elif len(args.key) == 0:
-                logging.warning("Warning: No keys specified.")
-    except Remote.ConnectionClosed:
+        if config["method"] == "legacy":
+            remote = RemoteLegacy(config)
+        elif config["method"] == "websocket":
+            remote = RemoteWebsocket(config)
```
Comment (Ape) [inline samsungctl/__main__.py]: Fail here with an error message if the method is something else.

E.g. "Error: Unknown method 'asdf'".

File: samsungctl/remote_websocket.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import websocket
+
+from . import exceptions
+
+class RemoteWebsocket():
+    """Object for remote control connection."""
+
+    def __init__(self, config):
+        if not config["port"]:
+            config["port"] = 8001;
```
Comment (Ape) [inline samsungctl/remote_websocket.py]: No need for the `;`.

File: samsungctl/remote_websocket.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        """Make a new connection."""
+        self.connection = websocket.WebSocket()
+        self.connection.connect("ws://%s:%d/api/v2/channels/samsung.remote.control?name=%s" % (config["host"], config["port"], self._serialize_string(config["name"])))
```
Comment (Ape) [inline samsungctl/remote_websocket.py]: Prefer `format` over `%`. Also, this line is too long.

Perhaps something like this:
```python
URL_FORMAT = "ws://{}:{}/api/v2/channels/samsung.remote.control?name={}"

...

self.connection.connect(URL_FORMAT.format(config["host"], config["port"],
                                          self._serialize_string(config["name"])))
```

**Comment by Ape** [comment]
Additionally, the `README.md` should be updated.

- [ ] The intro / description should briefly mention the legacy TCP protocol and the new websocket protocol and how that relates to the TV models.
- [ ] `websocket-client` should be added to Dependencies, but it should be optional.
- [ ] Command line usage should mention `--method`.
- [ ] Library section should be updated for RemoteLegacy and RemoteWebsocket.
- [ ] The example program needs some updates.

File: samsungctl/remote_websocket.py (hunk 2 above)
Comment (nunofgs) [inline samsungctl/remote_websocket.py]: Habit 😄

File: samsungctl/remote_websocket.py (hunk 3 above)
Comment (nunofgs) [inline samsungctl/remote_websocket.py]: Great suggestion! Done.

File: samsungctl/remote_websocket.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import base64
+import json
+import logging
+import socket
+import time
+import websocket
```
Comment (nunofgs) [inline samsungctl/remote_websocket.py]: I updated the PR with a slightly different approach.

The `Remote` class can now be used just as before. This avoids the obvious breaking change and keeps the public "api" of this module simple for any dependees, for example [home-assistant](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/samsungtv.py#L64).

I also moved the import to the _RemoteWebsocket_'s `__init__` method. This way, if someone specifies the _websocket_ method, we assume that the dependency was installed.

What do you think?

**Comment by nunofgs** [comment]
@kylerw since you demonstrated interest in the KS8000 support, can you give a hand in testing this?

File: samsungctl/remote_websocket.py (hunk 4 above)
Comment (Ape) [inline samsungctl/remote_websocket.py]: Thanks! I like your approach on this.

File: samsungctl/remote.py [hunk 5]
```diff
@@ -3,109 +3,21 @@
 import socket
 import time
```
Comment (Ape) [inline samsungctl/remote.py]: Could you remove all unnecessary imports?

**Comment by Ape** [comment]
Now that you have `Remote` class again, I was thinking if the implementations should inherit it (as they implement the same interface). Seems like `close()` implementation could be shared. The exceptions could also be on `Remote`.

This is just some thoughts. I don't have strong opinions on this.

**Comment by Ape** [comment]
Can you check that connecting to missing or otherwise invalid hosts will produce meaningful error messages? E.g. "Error: Connection refused" or "Error: Timed out!".

**Comment by Ape** [comment]
Since we already have References section in `README.md` perhaps you should add this line:

> https://github.com/kyleaa/homebridge-samsungtv2016 ('websocket' method)

**Comment by kylerw** [comment]
This is awesome @nunofgs!!! 

@Ape - I'll try fire it up tonight to test if you still need it.


I haven't dug into the changes, but does the update use the port in the config to pick identify websocket vs traditional automatically if the new parameter isn't identified (since the port was there before)? I believe this would make it easy for HomeAssistant and others to allow functionality with 2016+ models without much (if any) change required.

**Comment by kylerw** [comment]
Sorry guys, the week has gotten away from me. I'll work to test this this weekend and report back.

File: samsungctl/remote.py (hunk 5 above)
Comment (nunofgs) [inline samsungctl/remote.py]: Oops, will do.

**Comment by nunofgs** [comment]
> @Ape: Now that you have Remote class again, I was thinking if the implementations should inherit it (as they implement the same interface). Seems like close() implementation could be shared. The exceptions could also be on Remote.

Not sure if I understand what you mean here. If we make _RemoteLegacy_ and _RemoteWebsocket_ inherit from Remote, then we'd be forcing a BC since dependees would have to import the _RemoteLegacy_ class explicitly.

Am I looking at this wrong?


> @Ape: Can you check that connecting to missing or otherwise invalid hosts will produce meaningful error messages? E.g. "Error: Connection refused" or "Error: Timed out!".

Here are a couple of examples. The remaining errors are harder to simulate (unless we add unit tests) but they should be displayed in the same way.

```bash
❯ python3 -m samsungctl --host 192.168.1.92 --method websocket2 KEY_VOLDOWN
Error: Unknown method 'websocket2'

❯ python3 -m samsungctl --host 192.168.1.92 --method websocket KEY_VOLDOWN
Error: Connection refused
```

> @Ape: Since we already have References section in README.md perhaps you should add this line:

Great idea! Done.

> @kylerw: I haven't dug into the changes, but does the update use the port in the config to pick identify websocket vs traditional automatically if the new parameter isn't identified (since the port was there before)? I believe this would make it easy for HomeAssistant and others to allow functionality with 2016+ models without much (if any) change required.

I thought of this originally. It would indeed "automatically" add support for my Samsung KS8000 in Home Assistant without requiring them to do any code changes at all (which is pretty cool), but it felt a little strange to infer the method from the port.

If we go that route, then we'd have to assume that port 8001 is "websocket" and anything else is "legacy".

I'm fine with either implementation and will happily update the PR as needed. @Ape, any thoughts on this?

**Comment by Mofef** [comment]
using @nunofgs' code and `--method websocket` the program terminates without error, but the TV (UE40JU6500) does not react. With wireshark I found that the request

`    {"params": {"DataOfCmd": "KEY_MUTE", "Option": "false", "Cmd": "Click", "TypeOfRemote": "SendRemoteKey"}, "method": "ms.remote.control"}`

is answered by the TV with 

`{"event":"ms.error","data":{"message":"unrecognized method value : ms.remote.control"}}`

:/
Any ideas?

**Comment by nunofgs** [comment]
@Mofef: hmm interesting. I may have an idea. Can you install the Samsung [app](https://appsto.re/pt/x9fkZ.i) on your phone, pair it with your TV, and then try my PR again?

**Comment by Mofef** [comment]
@nunofgs I installed the Android version of it and paired it. but sadly the eror is stil the same. :/
Btw: Thanks for pointing out this app I didn't find the official Remote app from samsung so far. :)

**Comment by Mofef** [comment]
@nunofgs do you know how this app actually works? i wasn't able to trace any packages from it via wireshark oO. I also made sure to block the infrared port on my phone, so it can't use it

**Comment by nunofgs** [comment]
I admit I haven't sniffed the traffic to/from the app. I simply ported the code in https://github.com/kyleaa/homebridge-samsungtv2016.

Hmm, I'm stumped. I'll have to whip out wireshark as well and try to find the cause.

By the way, this is working flawlessly on my KS8000 TV. Here's a quick video:

img_6193

(I swear I'm not using the remote in the background haha 😄 )

@Mofef what TV model do you have?

**Comment by Mofef** [comment]
@nunofgs it's a UE40JU6550

**Comment by kylerw** [comment]
Finally got around to testing on my KS8000 using the command line and it is working like a charm.

**Comment by kylerw** [comment]
Another thought I had, would it make sense to add the wake on lan ability like (https://github.com/kyleaa/homebridge-samsungtv2016) implemented to allow turning the TV on?

**Comment by Ape** [comment]
I think we shouldn't modify the key codes. But, we could add new alias key codes that are translated properly to each device. E.g. `power` => `KEY_POWER` or `KEY_POWEROFF`. This would be a separate pull request.

I think the wake on lan capability is a bit out of scope for samsungctl. There are already utilities and libraries that do that just fine. It's not specific to Samsung TVs.

**Comment by Ape** [comment]
I think this pull request is soon ready to be merged. I'll make a new release, maybe 1.0.0, after a brief testing period after the merge.

But I'd like to have completely updated README.md before I'll merge this. See https://github.com/Ape/samsungctl/pull/19#issuecomment-260258791.

**Comment by Mofef** [comment]
Btw in the meanwhile I found that I can start the authentication process by sending "pin4" via POST to http://192.168.178.76:8080/ws/apps/CloudPINPage then the tv shows a 4 digit screen. Anyone who has a similarly behaving TV? Further experiences?

**Comment by nunofgs** [comment]
> @Ape But I'd like to have completely updated README.md before I'll merge this. See #19 (comment).

Oops, totally lost track of this. I just updated the PR with the changes you requested.

Also, I'm starting to wonder if the `method` parameter is really the way to go here. It requires every dependee to introduce these changes.

WDYT of implementing the change that @kylerw mentioned [here](https://github.com/Ape/samsungctl/pull/19#issuecomment-260974358)?

File: README.md [hunk 6]
```diff
@@ -130,6 +134,7 @@ config = {
     "id": "",
     "host": "192.168.0.10",
     "port": 55000,
+		"method": "legacy",
```
Comment (Ape) [inline README.md]: This is indented using tabs instead of spaces.

**Comment by kylerw** [comment]
Hey @nunofgs and @Ape any update on this PR? Would love to get the changes merged into HomeAssistant!

**Comment by Ape** [comment]
I'm sorry for the delay. I have been busy lately. Thanks for pushing me to finally reviewing and testing this PR.

**Comment by reedlaw** [comment]
I get `Error: Connection refused`. I have a Samsung KUF30E (should be the same as KU6300). Anyone get this model to work?

**Comment by almcelmon** [comment]
@kylerw I'll help test once you merge into Home assistant. I'm testing a 2016 8500 that currently just shows  the remote in hass. :D Looking forward to it!

**Comment by Ape** [comment]
@kylerw I released 0.6.0 with this feature.

**Comment by onemico** [comment]
@Mofef did you ever get to the bottom of your issue? I am having very similar issues on my TV, which appears to have web socket ports open but is a 2015 TV as far as I am aware.  Have paired with Samsung remote (my MyTifi) without any issues, but don't get offered a PIN for this.

**Pull request merged**